cpp/src/arrow/filesystem/s3_internal.h

```diff
@@ -158,6 +170,13 @@ Status ErrorToStatus(const std::string& prefix, const std::string& operation,
   if (error_type == Aws::S3::S3Errors::UNKNOWN) {
     ss << " (HTTP status " << static_cast<int>(error.GetResponseCode()) << ")";
   }
+  const auto maybe_region = BucketRegionFromError(error);
+  if (maybe_region.has_value() && region.has_value()) {
```

Comment:
Only do this when a region was passed to start with? (so moving `region.has_value()` before `BucketRegionFromError`)

Reply:
https://github.com/apache/arrow/pull/14601/commits/d1b6471570d7fc1f82ca1a885c8d96ad664f528f